File: シューティング学習/list/enemy/触手.cpp
#include <math.h>

void MoveTentacle(
	float x[], float y[],  // 各パーツの座標
	int num_part,          // パーツの数
	float tx, float ty,    // 目標の座標
	float v,               // 先端が移動する速さ
	float limit,           // 先端と根元の最大距離
	int num_loop           // 計算の繰り返し回数
) {
	// 先端の座標と根元の座標
	float hx=x[0], hy=y[0];
	float rx=x[num_part-1], ry=y[num_part-1];

	// 目標の方向に先端を動かす
	hx+=(tx>=hx+v ? v : (tx<=hx-v ? -v : 0));
	hy+=(ty>=hy+v ? v : (ty<=hy-v ? -v : 0));

	// 先端の移動範囲を制限する：
	// 先端と根元との距離が限界値を超えていたら，
	// 移動可能範囲内に戻す。
	float dx=hx-rx, dy=hy-ry;
	float d=sqrt(dx*dx+dy*dy);
	if (d>limit) {
		hx=dx*limit/d+rx;
		hy=dy*limit/d+ry;
	}

	// 中間部分の座標を計算する：
	// 隣接部分の座標の平均をとる。
	// 計算は先端から根元に向かって行う。
	// 必要に応じて計算を複数回繰り返す。
	x[0]=hx; y[0]=hy;
	for (int l=0; l<num_loop; l++) {
		for (int i=1; i<num_part-1; i++) {
			x[i]=(x[i-1]+x[i+1])/2;
			y[i]=(y[i-1]+y[i+1])/2;
		}
	}
}

File: シューティング学習/list/enemy/java/触手.java
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

public class 触手 extends JFrame implements Runnable {

	// 触手
	static final int WIDTH=800, HEIGHT=600, R=10;
	static final int NUM_PART=30;
	float[] X, Y;
	float TX, TY;

	// メインルーチン
	public static void main(String[] args) {
		(new Thread(new 触手())).start();
	}
	
	// コンストラクタ
	public 触手() {
		setTitle("触手");
		setSize(WIDTH, HEIGHT);

		// 座標の初期化
		X=new float[NUM_PART];
		Y=new float[NUM_PART];
		for (int i=0; i<NUM_PART; i++) {
			X[i]=WIDTH/2-i*WIDTH/2/NUM_PART;
			Y[i]=HEIGHT/2;
		}

		// 描画コンポーネント
		Container cp=getContentPane();
		cp.setLayout(new BorderLayout());
		cp.add(new JComponent() {
			public void paint(Graphics gra) {
				setBackground(Color.white);
				gra.clearRect(0, 0, 触手.WIDTH, 触手.HEIGHT);
				gra.setColor(Color.black);
				for (int i=0; i<NUM_PART; i++) {
					gra.fillOval(
						(int)X[i]-R, (int)Y[i]-R, 
						R*2, R*2);
				}
			}
		}, BorderLayout.CENTER);

		// マウス
		addMouseMotionListener(new MouseMotionAdapter() {
			public void mouseMoved(MouseEvent e) {
				TX=e.getX();
				TY=e.getY();
			}
		});

		// ウィンドウを閉じる
		addWindowListener(new WindowAdapter() {
			public void windowClosing(WindowEvent e) {
				System.exit(0);
			}
		});

		setVisible(true);
	}
	
	// スレッド
	public void run() {
		while (true) {
			updatePos();
			repaint();
			try { Thread.sleep(20); } 
				catch (InterruptedException ie) { break; }
		}
	}

	// 触手を動かす
	void updatePos() {
		final float LIMIT=600, MIN_LIMIT=200;
		final int WARP=5;
		final float V=4;
		X[0]+=(TX>X[0]+V?V:(TX<X[0]-V?-V:0));
		Y[0]+=(TY>Y[0]+V?V:(TY<Y[0]-V?-V:0));
		
		float dx=X[0]-X[NUM_PART-1], dy=Y[0]-Y[NUM_PART-1];
		float d=(float)Math.sqrt(dx*dx+dy*dy);
		if (d>LIMIT) {
			X[0]=dx*LIMIT/d+X[NUM_PART-1];
			Y[0]=dy*LIMIT/d+Y[NUM_PART-1];
		}
		if (d<MIN_LIMIT) {
			X[0]=dx*MIN_LIMIT/d+X[NUM_PART-1];
			Y[0]=dy*MIN_LIMIT/d+Y[NUM_PART-1];
		}
		
		for (int w=0; w<WARP; w++) {
		for (int i=1; i<NUM_PART-1; i++) {
			X[i]=(X[i-1]+X[i+1])/2;
			Y[i]=(Y[i-1]+Y[i+1])/2;
		}
		}
	}
}



